File: Budgeting application.rb
class BudgetingApp

  def initialize
    puts "Enter username: "
    @name = gets.chomp
    @total = 0
    @expenses_hash = Hash.new
    get_total
    menu

  end

  def get_total
  	print "Current budget: £"
  	@total = gets.chomp.to_i
  end
  
  def menu
    while true
      puts "\n"
      puts "Welcome #{@name}"
      puts "__________________________________________"
      puts "Type the number of the option you want"
      puts "__________________________________________"
      puts "1) Add expenses"
      puts "2) Update expenses"
      puts "3) Display current expenses"
      puts "4) Display remaining total"
      puts "5) Exit"
      puts "__________________________________________"
      print ": "

      choice = gets.chomp
      case choice
        when "1"
          add_expenses
        when "2"
          update_expense
        when "3"
          puts @expenses_hash
        when "4"
          puts "Current total: #{@total}\n"
        when "5"
          break
      end
    end
  end

  def repeat
    puts 'Repeat?: "Y" or "N"'
    result = gets.chomp.upcase
    if result[0] == "N"
      return true
    else
      return false
    end
  end

  def add_expenses

    flag = true
    while flag
      print "Name of expense: "
      expense = gets.chomp.to_sym
      print "Value of expense: £"
      value = gets.chomp.to_i
      @total -= value
      @expenses_hash[expense] = value

      flag = false if repeat
    end
    puts "New total: £#{@total}"
  end

  def update_expense

    flag = true
    while flag
      print "Expense to update: "
      search = gets.chomp.to_sym
      if @expenses_hash[search] != nil
        print "New value: £"
        new_value = gets.chomp.to_i
        old_value = @expenses_hash[search]
        diff = (old_value - new_value).abs
        if new_value > old_value
          @total -= diff
        else
          @total += diff
        end
        @expenses_hash[search] = new_value
        puts "Expense updated"
        puts "New total: £#{@total}"
      else 
        puts "Error: Expense not found"
      end
      flag = false if repeat
    end
  end
end

app = BudgetingApp.new

File: Budgeting application.php
<?php

/**
 * Author: Simon Addicott
 * Date: 12/12/2022
 * 
 */

class BudgetingApp {

  private $total;
  private $expensesHash;
  private $userName;

  function __construct()
  {
    $this->login();
    $this->total = 0;
    $this->expensesHash = [];
    $this->getTotal();
    $this->menu();
  }

  function login()
  {
    system("clear");
    $this->userName = readline("Enter username: ");

    $this->expensesHash[$this->user_name] = $this->userName;
    
    $this->password = readLine("Enter password: ");
    
  }

  public function getTotal()
  {
    system("clear");
  	$this->total = readline("Current budget: £");
  }
  
  public function menu()
  {
    system("clear");
    while(true){
      system("clear");
      echo "\n";
      echo "Welcome " . $this->userName . "\n";
      echo "__________________________________________\n";
      echo "Type the number of the option you want\n";
      echo "__________________________________________\n";
      echo "1) Add expenses\n";
      echo "2) Update expenses\n";
      echo "3) Display current expenses\n";
      echo "4) Display remaining total\n";
      echo "5) Exit\n";
      echo "__________________________________________\n";

      $choice = readline(": ");
      switch($choice){
        case "1":
          $this->addExpenses();
          break;
        case "2":
          $this->updateExpense();
          break;
        case "3":
          system("clear");
          print_r($this->expensesHash) . "\n";
          readline("Press enter to continue...");
          break;
        case "4":
          system("clear");
          echo "Current total: £" . $this->total . "\n";
          readline("Press enter to continue...");
          break;
        case "5":
          break;
      }
    }
  }

  private function repeat()
  {
    echo "\n\n";
    $result = strtoupper(readLine('Repeat?: "Y" or "N"'));
    
    if ($result == "N")
      return true;
    
    return false;
  }

  private function addExpenses()
  {
    system("clear");
    $flag = true;
    while($flag){
      $expense = readline("Name of expense: ");
      $value = readline("Value of expense: £");
      $this->total -= $value;
      $this->expensesHash[$this->userName][$expense] = $value;

      if($this->repeat()){
        $flag = false;
      }

    }
    echo "New total: £" . $this->total . "\n";
  }

  private function updateExpense() 
  {
    system("clear");
    $flagg = true;
    while($flagg)
    {
      $search = readline("Expense to update: ");

      if(isset($this->expenses_hash[$search])){

        $newValue =  readline("New value: £");
        $oldValue = $this->expensesHash[$search];
        $diff = abs(($oldValue - $newValue));
        if ($newValue > $oldValue)
          $this->total -= $diff;
        else
          $this->total += $diff;
        
        $this->expenses_hash[$search] = $newValue;

        echo "Expense updated\n";
        echo  "New total: £" . $this->total . "\n";
      } else { 
        echo  "Error: Expense not found\n";
      }

      if($this->repeat()){
        $flagg = false;
      }

    }
  }
}

$app = new BudgetingApp();

?>